src/PrestaShopBundle/Entity/Repository/LogRepository.php

```diff
@@ -37,6 +38,8 @@
  */
 class LogRepository implements RepositoryInterface, DoctrineQueryBuilderInterface
 {
+    use QueryBuilderHelperTrait;
```

Comment:
As this function you introduce will be only used in repositories, you'd better use an abstract class.

Reply:
ping @eternoendless, wdyt? How do you decide when to use an abstract class or a trait?

Reply:
in this case, trait does not provide any public API and it does not have any dependencies. So you can use it in both repositories or other services responsible for building queries. 

p.s. @mickaelandrieu i know you dont like traits. :smiley:

Reply:
It's not about like or not like, but I'm not sure we should use this traits "outside" of repositories.

For instance, I'm easy with something like TranslatorTrait as I always change the signature of `trans` function to make domain the second argument. As translations are needed everywhere and on a lot of unrelated classes, I guess this is a valid use case for a trait.

I'd like the opinion of @eternoendless or @PierreRambaud or anyone form @PrestaShop/prestashop-core-dev here :+1:

Reply:
not exactly to the topic, but i did use query building outside for repository here https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Grid/Query/EmailLogsQueryBuilder.php do you think it should be updated to repository?

Reply:
I'm not a fan of traits. Some define them as "language-assisted copy-paste" – which in all fairness is pretty much true.

While they may seem useful at first sight, traits are really treacherous. They can be too easily misused to violate the Single Responsibility principle, because they provide the _illusion_ of clean design and composition. But make no mistake, when you add a trait, you are actually _appending_ behavior to your class. It just happens to be in another file, hidden from view.

So whenever you are tempted to add a trait, ask yourself: "are the methods in this trait part of the class' responsibility?" Most of the time they will not, so you can get it done using real composition (ie. injecting another object) instead.

Reply:
@eternoendless in this case method provided by trait is this class responsibility. This method does not provide public API for class.

Reason why i added this trait is that i noticed that all query builder have exactly the same few lines of code (adding sorting & pagination to doctrine queries). 

what do you think?

Reply:
This is reusable logic that I think it would make more sense as a separate class. To me, it looks like it enriches a QueryBuilder using SearchCriteriaInterface. As a bonus, you will be able to add unit test it (try it!).

Think of it as a class whose role is to enrich a query (any query) using SearchCriteria. 

Usage would be like this:

```php
$this->getSearchCriteriaApplicator()
    ->applySorting($searchCriteria, $qb)
    ->applyPagination($searchCriteria, $qb);
```

Following this train of though, I think the `buildGridQuery` method could be even further enhanced.

Check out this block:

```php
$filters = $searchCriteria->getFilters();
foreach ($filters as $filterName => $filterValue) {
    if (empty($filterValue)) {
        continue;
    }

    if ('employee' == $filterName) {
        $qb->andWhere('e.lastname LIKE :employee OR e.firstname LIKE :employee');
        $qb->setParameter('employee', '%'.$filterValue.'%');
        continue;
    }

    if ('date_add' == $filterName) {
        if (!empty($filterValue['from']) &&
            !empty($filterValue['to'])
        ) {
            $qb->andWhere('l.date_add BETWEEN :date_from AND :date_to');
            $qb->setParameters(array(
                'date_from' => $filterValue['from'],
                'date_to' => $filterValue['to'],
            ));
        }
        continue;
    }

    $qb->andWhere("$filterName LIKE :$filterName");
    $qb->setParameter($filterName, '%'.$filterValue.'%');
}
```

There are two problems with it.

First, [Switch Is Bad](https://sourcemaking.com/refactoring/smells/switch-statements). Same goes for a sequence of `if`. Think about this: How would you add another filter? By editing the file? How would _somebody else_ do it?.

Second, this is a typical example of procedural micro-management, as opposed to the ["Tell, don't ask" principle](https://pragprog.com/articles/tell-dont-ask):

> Procedural code gets information then makes decisions. Object-oriented code tells objects to do things. 
> — Alec Sharp

The first thing to do is identifying our actors:

- We have a `SearchCriteriaInterface` which defines a couple known criteria (`orderBy`, `orderWay`, `offset` and `limit`) as well a number of unknown criteria (`filters`).
- Then there is a `QueryBuilder` where we need to apply that criteria.

Currently, the knowledge on how to match that `SearchCriteria` to the `QueryBuilder` is hardcoded in the `Repository`.

This can be extracted using a `SearchCriteriaApplicator`, and replace the whole block with this:

```php
$this->getSearchCriteriaApplicator()->apply($searchCriteria, $qb);
```

Now that looks simpler! We simply extracted the previous code and put it inside a `SearchCriteriaApplicator::apply()` method. This solves the "tell, don't ask" issue and makes code easier to understand.

But we still have to ugly, hard-to-extend sequence of `if` to refactor. How can we improve on that?

First, let's extract each `if` case into its own method:

```php
// class SearchCriteriaApplicator

public function apply(SearchCriteriaInterface $searchCriteria, QueryBuilder $qb)
{
    $filters = $searchCriteria->getFilters();
    foreach ($filters as $filterName => $filterValue) {
        if (empty($filterValue)) {
            continue;
        }
    
        if ($this->applyEmployee($filterName, $filterValue $qb)) {
            continue;
        }
        
        if ($this->applyDateAdd($filterName, $filterValue, $qb)) {
            continue;
        }
    
        $this->applyOther($filterName, $filterValue, $qb);
    }
}

// [...]

private function applyEmployee($filterName, $filterValue, QueryBuilder $qb)
{
    if ('employee' == $filterName) {
        $qb->andWhere('e.lastname LIKE :employee OR e.firstname LIKE :employee');
        $qb->setParameter('employee', '%'.$filterValue.'%');
        return true;
    }
    return false;
}

private function applyDateAdd($filterName, $filterValue, QueryBuilder $qb)
{
    if ('date_add' == $filterName) {
        if (!empty($filterValue['from']) &&
            !empty($filterValue['to'])
        ) {
            $qb->andWhere('l.date_add BETWEEN :date_from AND :date_to');
            $qb->setParameters(array(
                'date_from' => $filterValue['from'],
                'date_to' => $filterValue['to'],
            ));
        }
        return true;
    }
    return false;
}

// [...]
```

Now, if we refactor it a little further and extract previously extracted methods into classes of their own, we can stop looping on the the `$filters` and loop on `FilterApplicators` instead:

```php
// class SearchCriteriaApplicator

/** @param FilterApplicatorInterface[] $criteriaApplicator */
public function __construct(array $criteriaApplicators);

public function apply(SearchCriteriaInterface $searchCriteria, QueryBuilder $qb)
{
    $filters = $searchCriteria->getFilters();
    if (!empty($filters) {    
        foreach ($this->getCriteriaApplicators() as $applicator) {
            $applicator->apply($filters, $qb);
        }
    }
}

// class EmployeeApplicator

private function match($filters)
{
    return isset($filters['employee']) && !empty($filters['employee']);
}

public function apply(array $filters, QueryBuilder $qb)
{
    if ($this->match($filters) {
        $qb->andWhere('e.lastname LIKE :employee OR e.firstname LIKE :employee');
        $qb->setParameter('employee', '%'.$filter['employee'].'%');
    }
}

// class DateAddApplicator

private function match($filters)
{
    if (!isset($filters['date_add']) {
        return false;
    }
    $filterValue = $filters['date_add'];
    
    return !empty($filterValue['from']) && !empty($filterValue['to']);
}

public function apply(array $filters, QueryBuilder $qb)
{
    if ($this->match($filters) {
        $filterValue = $filters['date_add'];
        
        $qb->andWhere('l.date_add BETWEEN :date_from AND :date_to');
        $qb->setParameters(array(
            'date_from' => $filterValue['from'],
            'date_to' => $filterValue['to'],
        ));
    }
    return true;
}
```

This way, we are inversing responsibility, and making applicators _subscribe_ to the presence of a filter on a filter list instead of calling them explicitely. To add a new filter, you only need to add a new `FilterApplicator` on the list contained by `SearchCriteriaApplicator`, which can of course be done in runtime (no more file editing!).

Reply:
As I totally agree with what you're saying, I have the feeling the refactoring you are suggesting should be done in another pull request to make the scope of this one limited.

How about creating an issue into prestafony project and set a milestone to ensure this will be done during 1.7.5 "time of development" ? :+1:

Reply:
@eternoendless as usual, great explanation and i really like your idea!

How would you suggest dealing with aliases in SQL? For example:

```php
public function apply(array $filters, QueryBuilder $qb)
{
    if ($this->match($filters) {
        $filterValue = $filters['date_add'];
        
        $qb->andWhere('l.date_add BETWEEN :date_from AND :date_to');
        $qb->setParameters(array(
            'date_from' => $filterValue['from'],
            'date_to' => $filterValue['to'],
        ));
    }
    return true;
}
```
And in this line ` $qb->andWhere('l.date_add BETWEEN :date_from AND :date_to');` we have `l.date_add`,  but in other SQL it may be `a.date_add` or `b.date_add`.

Would we need different applicators for this? Or maybe we should try to avoid alias and just keep ` $qb->andWhere('date_add BETWEEN :date_from AND :date_to');` so applicator would be reusable?

Reply:
@mickaelandrieu 
> I have the feeling the refactoring you are suggesting should be done in another pull request

Absolutely. The second-level refactoring (which extracts code from `buidGridQuery()`) can be done in a separate PR.

However, the first one (`$this->getSearchCriteriaApplicator()->applySorting($searchCriteria, $qb)->applyPagination($searchCriteria, $qb)`) I'd like it to be done in this PR so that we can get rid of the trait.

@sarjon 
> How would you suggest dealing with aliases in SQL?

That's a very good question. I can think of two ways of solving it: either architecture-based or conventions-based.

In the architecture-based solution, we would have to think of a way of providing Applicators with meta information about how tables are aliased. This information would be provided on construction and used to match a known table to the alias it happens to have in the query builder. While this would may solve our immediate problem, it would also complexity to the whole system.

After a closer look, it looks like it would also be possible to figure out the alias by inspecting the query builder:

```php
$qb->getPart('from')
// contains a collection of array('table' => $tableName, 'alias' => $alias')
```

Of course, also "smells" because of added complexity. Also, aliases may be defined by `JOIN` which would add further complexity.

In the conventions-based solution, we accept that Applicators are closely coupled to the queries they change, and as such they may share implicit (aka. "hard-coded") knowledge. This will probably mean having two Applicators that do pretty much the same thing but with slightly different code because of aliases. If you worry about aliases, then use full table names everywhere.

No solution is perfect, both have tradeoffs.